lambda-buffers-proto-compat/src/LambdaBuffers/ProtoCompat/ProtoCompat.hs

```diff
@@ -0,0 +1,562 @@
+
+data TyRef = TyRef {
+  tyRef :: TyRefInner,
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data TyDef = TyDef {
+  tyName     :: TyName,
+  tyArgs     :: [TyArg],
+  tyBody     :: TyBody,
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data TyArg = TyArg {
+ argName :: ArgName,
+ argKind :: Kind,
+ sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data TyBodyInner
+  = OpaqueI SourceInfo
+  | SumI Sum
+  deriving (Show, Eq, Ord, Generic)
+
+data TyBody = TyBody {
+  tyBody :: TyBodyInner,
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data Constructor = Constructor {
+  constrName :: ConstrName,
+  product    :: Product
+} deriving (Show, Eq, Ord, Generic)
+
+data Sum = Sum {
+  constructors :: NonEmpty Constructor,
+  sourceInfo   :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data Field = Field {
+  fieldName :: FieldName,
+  fieldTy   :: Ty
+} deriving (Show, Eq, Ord, Generic)
+
+data ProductInner
+  = EmptyI
+  | RecordI (NonEmpty Field)
+  | TupleI [Ty]
+  deriving (Show, Eq, Ord, Generic)
+
+data Product = Product {
+  product :: ProductInner,
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data ClassDef = ClassDef {
+  className     :: ClassName,
+  classArgs     :: TyArg, -- n.b. MPTC support is complicated and likely can't be implemented for V1
+  supers        :: [Constraint],
+  documentation :: Text,
+  sourceInfo    :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data InstanceClause = InstanceClause {
+  className   :: ClassName,
+  head        :: Ty,
+  constraints :: [Constraint],
+  sourceInfo  :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data Constraint = Constraint {
+  className :: ClassName,
+  argument  :: Ty, -- n.b. see previous n.b., for now constraints are monadic
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+data Module = Module {
+  moduleName :: ModuleName,
+  typeDefs   :: [TyDef],
+  classDefs  :: [ClassDef],
+  instances  :: [InstanceClause],
+  sourceInfo :: SourceInfo
+} deriving (Show, Eq, Ord, Generic)
+
+newtype CompilerInput = CompilerInput {modules :: [Module]}
+  deriving (Show, Eq, Ord, Generic)
+
+data FromProtoErr
+  = MultipleInstanceHeads ClassName [Ty] SourceInfo
+  | NoInstanceHead ClassName SourceInfo
+  | NoConstraintArgs ClassName SourceInfo
+  | MultipleConstraintArgs ClassName [Ty] SourceInfo
+  | NoClassArgs ClassName SourceInfo
+  | MultipleClassArgs ClassName SourceInfo
+  | NoTyAppArgs SourceInfo
+  | EmptyRecordBody SourceInfo -- Ideally we should catch & rethrow this and the next one when we have access to the tyname
+  | EmptySumBody SourceInfo
+  | EmptyName SourceInfo
+  | NegativeArity ArgName SourceInfo
+  | OneOfNotSet Text SourceInfo -- catchall
+  deriving (Show, Eq, Ord, Generic)
+
+class IsMessage (proto :: Type) (good :: Type) where
+  fromProto :: proto -> Either FromProtoErr good
+
+  toProto   :: good -> proto
```

Comment:
Do we really need a typeclass for this?

Reply:
Well it's either that or we have a billion functions & try to remember the name of each one. Typeclass is simpler.